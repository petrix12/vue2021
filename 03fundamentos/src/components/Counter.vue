<template>
    <!-- <h2>{{ title || 'Contador' }}</h2> -->
    <h2>{{ customTitle }}</h2>
    <p>{{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
    <!-- <p>{{ counter }}</p> -->
    <p data-testid="counter">{{ counter }}</p>
    <div class="buttons-container">
        <!-- <button v-on:click="increase">+1</button>
        <button v-on:click="decrease">-1</button> -->
        <!-- Las siguientes dos líneas de código son equivalentes a las dos comentadas anteriormente -->
        <button @click="increase">+1</button>
        <button @click="decrease">-1</button>
    </div>
</template>

<script>
export default {
    /* props: ['title', 'start'], */
    props: {
        title: String, 
        start: {
            type: Number,
            // required: true
            default: 100,
            validator1(value){
                return value >= 0
            }
        }
    },
    data(){
        return {
            counter: this.start
        }
    },
    methods: {
        getSquareValue(){
            return this.counter * this.counter
        },
        increase(){
            this.counter++
        },
        decrease(){
            this.counter--
        }
    },
    computed: {
        squareCounter(){
            return this.counter * this.counter
        },
        customTitle(){
            /* console.log(this.title)
            return this.title */
            return this.title || 'Contador'
        }
    }
}
</script>

<style>
button{
    background-color: #64BB87;
    border-radius: 5px;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
}

button:hover{
    background-color: #5aa67b;
    transition: 0.3s ease-in-out;
}
</style>