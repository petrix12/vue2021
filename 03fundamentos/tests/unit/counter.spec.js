import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter'

describe('Counter Component', () => {
    let wrapper
    beforeEach(() => {
        wrapper = shallowMount(Counter)
    })

    /* test('debe de hacer match con el snapshot', () => {
        const wrapper = shallowMount(Counter)
        expect( wrapper.html() ).toMatchSnapshot()
    }) */
    
    test('h2 debe de tener el valor por defecto "Contador"', () => {
        /* const h2 = wrapper.find('h2') */
        /* console.log(h2.text()) */
        /* expect( h2.text() ).toBe('Contador!') */
        expect(wrapper.find('h2').exists()).toBeTruthy()
        const h2Value = wrapper.find('h2').text()
        expect(h2Value).toBe('Contador')
    })

    test('El valor por defecto debe de ser 100 en el p', async() => {
        /* const pTags = wrapper.findAll('p') */
        const value = wrapper.find('[data-testid="counter"]').text()
        /* console.log(pTags) */
        /* expect(pTags[1].text()).toBe('100') */
        expect(value).toBe('100')

        const [increaseBtn, decreaseBtn] = wrapper.findAll('button')
        await increaseBtn.trigger('click')
    })

    test('Debe de incrementer en 1 el valor del contador', async() => {
        const [increaseBtn, decreaseBtn] = wrapper.findAll('button')

        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')
        await increaseBtn.trigger('click')
        await decreaseBtn.trigger('click')
        await decreaseBtn.trigger('click')

        const value = wrapper.find('[data-testid="counter"]').text()
        expect(value).toBe('101')
    })

    test('Debe de establecer el valor por defecto', () => {
        /* console.log(wrapper.props()) */
        /* const start = wrapper.props('start') */
        // La siguiente instrucci√≥n es equivalente a la anteriormente comentada
        const {start} = wrapper.props()
        /* console.log(start, typeof start) */

        const value = wrapper.find('[data-testid="counter"]').text()
        expect(Number(value)).toBe(start)
    })
})