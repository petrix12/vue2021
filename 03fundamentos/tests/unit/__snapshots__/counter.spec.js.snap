// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component debe de hacer match con el snapshot 1`] = `
<!-- <h2>{{ title || 'Contador' }}</h2> -->
<h2>Contador</h2>
<p>100 <sup>2</sup> = 10000</p>
<div>
  <!-- <button v-on:click="increase">+1</button>
        <button v-on:click="decrease">-1</button> -->
  <!-- Las siguientes dos líneas de código son equivalentes a las dos comentadas anteriormente --><button>+1</button><button>-1</button>
</div>
`;
